### PR TITLE
vmm: decouple legacy dev init from IRQCHIP creation

### PR DESCRIPTION
**Pull request by acatangiu**
Register legacy devices IRQs as part of legacy device registration, do IRQCHIP and PIT creation independently as part of (kvm) VM init.

Also do a bit of refactoring around kernel cmdline to move some ugly logic out of vmm.

Setting up an interrupt controller should, well.. set up an interrupt controller (irqchip) without having a dependency on a particular list of devices that will eventually use it.

Move the registering of legacy devs irqs to `attach_legacy_devices()` which already does the registering of legacy devs bus addrs.

Before this patch the implementation was not just spaghetti but also incorrect because it would register irqs for devices not yet attached.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by andreeaflorescu** [comment]
@acatangiu can you please add something about the "why" you decided to do this in your commit message? Specifically, why is it better to separate the creation of irqchip from irq events?

**Comment by andreeaflorescu** [comment]
@acatangiu thanks for adding the explanation. Honestly, I didn't understand what was the problem with the initial code. Now it makes sense! :-1:

**Comment by acatangiu** [comment]
@andreeaflorescu You're right, it was clear to me because I was working in that area, but it's indeed better to clearly explain it in the commit message.

File: kernel/src/cmdline/mod.rs [hunk 1]
```diff
@@ -140,32 +157,33 @@ impl Cmdline {
...
     pub fn as_str(&self) -> &str {
         self.line.as_str()
     }
-}
 
-impl Into<Vec<u8>> for Cmdline {
-    fn into(self) -> Vec<u8> {
-        self.line.into_bytes()
+    /// Returns the cmdline in progress as CString.
+    pub fn as_cstring(&self) -> Result<CString> {
```
Comment (dianpopa) [inline kernel/src/cmdline/mod.rs]: 👍

File: kernel/src/cmdline/mod.rs [hunk 2]
```diff
@@ -29,6 +34,8 @@ impl fmt::Display for Error {
             f,
             "{}",
             match *self {
+                Error::CommandLineCopy => "Failed to copy the command line string to guest memory",
+                Error::CommandLineOverflow => "Command line string overflows guest memory",
                 Error::InvalidAscii => "string contains non-printable ASCII character",
```
Comment (dianpopa) [inline kernel/src/cmdline/mod.rs]: nit: Let s use same format for all enum values:
```suggestion
                Error::InvalidAscii => "Command line string contains non-printable ASCII character",
```

File: kernel/src/cmdline/mod.rs [hunk 3]
```diff
@@ -29,6 +34,8 @@ impl fmt::Display for Error {
...
                 Error::HasSpace => "string contains a space",
```
Comment (dianpopa) [inline kernel/src/cmdline/mod.rs]:
```suggestion
                Error::HasSpace => "Command line string contains a space",
```

File: kernel/src/cmdline/mod.rs [hunk 4]
```diff
@@ -29,6 +34,8 @@ impl fmt::Display for Error {
...
                 Error::HasEquals => "string contains an equals sign",
```
Comment (dianpopa) [inline kernel/src/cmdline/mod.rs]:
```suggestion
                Error::HasEquals => "Command line string contains an equals sign",
```

File: api_server/src/request/mod.rs [hunk 5]
```diff
@@ -324,12 +324,12 @@ mod tests {
         check_error_response(vmm_resp, StatusCode::BadRequest);
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply api_server/src/request/mod.rs]: Yes :laughing:, is it confusing?

[the remark this answers is not in the document]
Comment (acatangiu) [reply api_server/src/request/mod.rs]: Changed it to be more clear.

**Pull request merged**